mempool/reactor.go

```diff
@@ -145,13 +190,23 @@ func (memR *MempoolReactor) broadcastTxRoutine(peer p2p.Peer) {
 			time.Sleep(peerCatchupSleepIntervalMS * time.Millisecond)
 			continue
 		}
+		// ensure peer hasn't already sent us this tx
+		sendTxToPeer := true
+		for _, sender := range memTx.senders {
+			if peerID == sender {
+				sendTxToPeer = false
+				break
+			}
+		}
 
-		// send memTx
-		msg := &TxMessage{Tx: memTx.tx}
-		success := peer.Send(MempoolChannel, cdc.MustMarshalBinaryBare(msg))
-		if !success {
-			time.Sleep(peerCatchupSleepIntervalMS * time.Millisecond)
-			continue
+		if sendTxToPeer {
+			// send memTx
+			msg := &TxMessage{Tx: memTx.tx}
```

Comment:
What if we send an information about the sender of a transaction along with it? (rather then remembering it?)

![blank diagram](https://user-images.githubusercontent.com/1282182/51823022-7e22ff00-22f7-11e9-9080-3898cdfca41b.png)

Just an idea.

Reply:
I feel like this makes eclipsing another node too easy. (Just have every tx you gossip include the node to eclipse in its list)